File: python/linked_lists/linked_list.py
"""
A simple linked list.
"""

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def push(self, item):
        new_node = Node(item)
        new_node.next = self.head
        self.head = new_node

    @property
    def lenth(self):
        cur = self.head
        res = 0
        while cur:
            res += 1
            cur = cur.next
        return res

class LinkedListWithTail:
    def __init__(self):
        self.head = None
        self.tail = None

    def push(self, item):
        new_node = Node(item)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node

    def append(self, item):
        new_node = Node(item)
        if not self.tail:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node

File: javascript/linked_lists/linked_list.js
function LinkedlistNode(value) {
  this.value = value;
  this.next = null;
}

function LinkedList() {
  this.head = null;
}

LinkedList.prototype.add = function(value) {
  var newNode = new LinkedlistNode(value);
  newNode.next = this.head;
  this.head = newNode;
}

LinkedList.prototype.remove = function() {

  if (this.head) { // at least one node
    var removed = this.head;
    this.head = this.head.next;
    return removed.value;

  } else { // no nodes
    return null;
  }
}

LinkedList.prototype.contains = function(targetValue) {
  var curNode = this.head;
  while (curNode && curNode.value != targetValue) {
    curNode = curNode.next;
  }
  return !!curNode;
}

LinkedList.prototype.length = function() {
  var len = 0,
      cur = this.head;

  while (cur.value) {
    cur = cur.next;
    len++;
  }

  return len;
}

// driver
let myList = new LinkedList()
let values = [1,2,3,4,5,6,7,8,9];
values.forEach((v)=>{
  myList.add(v);
})
